### ISSUE TITLE
Resolve Pitest Issues - AbstractHeaderCheck (1)

### ISSUE THREAD

**Issue by Devansu-Yadav** [checkstyle/checkstyle#9987]
Child Issue of #7797 ,
https://github.com/checkstyle/checkstyle/blob/master/.ci/pitest.sh#L54

**Comment by Devansu-Yadav** [comment]
This issue was opened as a GSoC 3rd issue. @romani @rnveach Could you please approve it at your convenience?

**Comment by Devansu-Yadav** [comment]
I am working on it.

**Comment by Devansu-Yadav** [comment]
### Pitest Report
https://devansu-yadav.github.io/Devansu-Yadav-Checkstyle-test-reports.github.io/issue9987/pit-reports/202105050706/com.puppycrawl.tools.checkstyle.checks.header/AbstractHeaderCheck.java.html

### Surviving mutation on line 156:
`removed call to java/lang/IllegalArgumentException::<init> → NO_COVERAGE`

Hardcoded mutation is [here](https://github.com/Devansu-Yadav/checkstyle/commit/d61c666c0c6049332763d9f700a7830397d9a2b2).

I will be updating this comment to add the diff regression report as well as the config.

**Comment by Devansu-Yadav** [comment]
I ran
```
groovy diff.groovy --localGitRepo /home/devansu/Java_Development/checkstyle --baseBranch fix_#9987 --patchBranch issue_9987_mutated_branch --config /home/devansu/Java_Development/config.xml --listOfProjects projects-to-test-on.properties
```
to generate diff report, but I am getting errors.

here's a gist [https://gist.github.com/Devansu-Yadav/f4b2a8140a3b33123635d6efdceea46a](https://gist.github.com/Devansu-Yadav/f4b2a8140a3b33123635d6efdceea46a) that consists of the `config.xml` that I have used and the **errors** that I am getting when regression runs on "**sevntu checkstyle**" project. After that the process just stops.

I checked my command again and it seems to be correct, any help would be appreciated.
@nmancus1 @AkMo3 

I think it could be because of the `config` that I am using where for property `headerFile` I have set to `config/java,header` which might not be there in some project in the projects list(`projects-to-test-on.properties`) but I am not sure about this.

**Comment by rnveach** [comment]
https://gist.github.com/Devansu-Yadav/f4b2a8140a3b33123635d6efdceea46a#file-diffgroovy_errors-txt-L507
> Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to find: config/java.header

https://gist.github.com/Devansu-Yadav/f4b2a8140a3b33123635d6efdceea46a#file-config_for_regression-xml-L28

Use absolute file locations for these `java.header` files you are giving it.

This check is special compared to others in that you not only have to create unique check configurations, but you also need to create unique input files to be used. Your input files will just be as important to for regression as the check configurations. You will want to write them specifically to target the areas pitest is pointing to.

https://github.com/checkstyle/checkstyle/blob/master/.ci/pitest.sh#L54
> `throw new IllegalArgumentException(&#34;unable to load header&#34;, ex);`

It looks like you may not be able to use regression in this case since you are trying to hit a throw statement. This throw statement is only hit when we have to abort everything. Regression is more for cases that either print or don't print violations.

You will have to advance to 4b.

**Comment by AkMo3** [comment]
I had earlier tried to resolve this one long back and these were my observation at that time. I don't remember correctly why I didn't make PR, I think I had low confidence that time :) , but this might help you to resolve this quicker

Here both can `catch` and `throw` expression
https://github.com/checkstyle/checkstyle/blob/96a383ba3cea293e29c1bde4abce8c6885abe827/.ci/pitest.sh#L53-L54
can be removed I think as:

This exception is being thrown because of calling of function `loadHeader()` at line 153
https://github.com/checkstyle/checkstyle/blob/96a383ba3cea293e29c1bde4abce8c6885abe827/src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java#L145-L159

The function `loadHeader()` will throw an exception when an `IOException` when `line = lnr.readLine();`  throws an exception https://github.com/checkstyle/checkstyle/blob/96a383ba3cea293e29c1bde4abce8c6885abe827/src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java#L171
`readLine()` throws an exception in cases like a sudden appearance of the null value or the file doesn't exist from where it was reading etc (I assume this should)

But in this cases exception would be either we not be thrown because we are checking for non-null value at start of `setHeader()` method
https://github.com/checkstyle/checkstyle/blob/96a383ba3cea293e29c1bde4abce8c6885abe827/src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java#L146
Even if it passes that two methods `matcher()`https://github.com/checkstyle/checkstyle/blob/96a383ba3cea293e29c1bde4abce8c6885abe827/src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java#L150 
and calling of `StringReader` class https://github.com/checkstyle/checkstyle/blob/96a383ba3cea293e29c1bde4abce8c6885abe827/src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java#L152

should throw an exception before `loadHeader()` is called.

@rnveach Sir, what do think if this?

**Comment by Devansu-Yadav** [comment]
@rnveach @romani After researching and thinking about different cases that could hit the `IOException`, I think that this exception won't occur in the `setHeader()` method. I think that we can safely remove the `try-catch` block and just instantiate the `Reader` object and pass it to `loadHeader()` method. 

Also, this issue will have to tackle both the mutations [https://github.com/checkstyle/checkstyle/blob/master/.ci/pitest.sh#L53-L54](https://github.com/checkstyle/checkstyle/blob/master/.ci/pitest.sh#L53-L54) if the mentors agree that the `catch` and `throw` expressions can be removed. 
Pls let me know what you think.

**Comment by romani** [comment]
@Devansu-Yadav , just send a PR, green CI and no regression in diff report will prove that fix is valid.

**Comment by pbludov** [comment]
Fixed by #10932
